Fix OnUserBeganPlaying not being invoked if already watching

smoogipoo opened the pull request:
When a user is watched via `WatchUser()`, the server sends out a `UserBeganPlaying` event. If the user is already being watched, `WatchUser()` is a client-side no-op.

If there's multiple places that bind to `OnUserBeganPlaying`, the second place will not receive a callback in this case.

smoogipoo committed:
Fix OnUserBeganPlaying not being invoked if already watching

peppy commented:
Not saying it's the direction we should go, but did you consider a `BindableList` for this instead?

smoogipoo commented:
I did, I wasn't sure if that was a good way to go because it removes the begin/finish naming and makes it so you have to rely on the actions performed on a list.

I'll try it out though, we have another bindable list (`PlayingUsers`) which is doing a similar thing...

smoogipoo commented:
It becomes something like this:
![image](https://user-images.githubusercontent.com/1329837/114990631-aba5f400-9ed3-11eb-9db6-1c12a06705a7.png)
Not sure I feel good about that...

peppy commented:
hmmmmmm, yeah. let's just go with what you have for now. the only reason i'm slightly against it is because it's a pattern we don't use anywhere else right now.

smoogipoo commented:
If you'd prefer, I can make a `BindOnUserBeganPlaying()` so at least it's similar to Bindable. Still needs to be unbound on dispose.

peppy commented:
May want to check on the test failures this had: https://ci.appveyor.com/project/peppy/osu/builds/38730515

smoogipoo commented:
Yep, I'm currently resolving them.

smoogipoo committed:
Change to explicit method instead

smoogipoo committed:
Fix test incorrect sending state too often

smoogipoo committed:
Remove now unnecessary (duplicating) test

smoogipoo committed:
Fix SpectatorScreen potentially missing user playing callbacks

smoogipoo commented:
Added the `SpectatorScreen` case which I originally intended to fix with this.

bdach commented:
Two things that spring to mind about this:

1. What happens if the watching user's connection drops and then goes back up again? I'm talking about this code path in particular:

	https://github.com/ppy/osu/blob/1832f624e20a1b0c5ca9365e0c31abc29be1a148/osu.Game/Online/Spectator/SpectatorStreamingClient.cs#L102-L128

	`currentUserStates` is not changed in this code path in this PR in any way, which means it's possibly storing stale data. On the surface looks like a potential desync issue to me.

2. `SpectatorScreen` has its own `spectatorStates` dictionary. Maybe it's not necessary any longer if the dictionary was exposed at a `SpectatorStreamingClient` level? It'd probably have to be exposed as immutable, though, so I'm not sure whether the excursion is worth it.

smoogipoo committed:
Clear current user states on disconnect

smoogipoo committed:
Lock around playingUsers

smoogipoo committed:
Rename currentUserStates -> playingUserStates

smoogipoo committed:
Remove local state dictionary from SpectatorScreen

smoogipoo committed:
Merge branch 'master' into fix-initial-spectator-state-callback

smoogipoo commented:
Good catch on (1) there. It's only the `false` path that needs to be changed - the `true` path goes through `WatchUser()` which is guaranteed to be a remote call via that code path. It's also revealed that `playingUsers` should also be inside the lock, because that code path is happening on a random TPL thread.

As for (2), we don't have a `BindableDictionary` yet, so I've exposed it via a `TryGetX` method from `SpectatorStreamingClient` itself.

bdach commented on osu.Game/Screens/Spectate/SpectatorScreen.cs:
```diff
@@ -62,26 +61,42 @@ protected override void LoadComplete()
-            spectatorClient.OnNewFrames += userSentFrames;
+            populateAllUsers().ContinueWith(_ => Schedule(() =>
+            {
+                spectatorClient.BindUserBeganPlaying(userBeganPlaying, true);
+                spectatorClient.OnUserFinishedPlaying += userFinishedPlaying;
+                spectatorClient.OnNewFrames += userSentFrames;
```
I didn't realise this before but this subscription happens quite a bit later than it did previously. I'm thinking there could be a possibility of a `OnUserBeganPlaying`/`OnUserFinishedPlaying`/`OnNewFrames` event firing before `populateAllUsers()` has completed, which would lead to those events getting dropped.

I guess this was moved because it doesn't make sense to perform the `BindUserBeganPlaying(..., true)` call before users have been populated? Maybe best to hoist it out of the continuation and drop the `true`?

peppy replied:
Are you talking specifically about `OnNewFrames` here? If so, I don't think dropping some frames is considered a huge deal (and given the code just above it would ensure we don't receiver frames until a `UserBeganPlaying` has fired, which I *think* is what we want/expect).

bdach replied:
You're right that dropping frames is not a huge deal, but dropping a starting/finishing play event might be. Especially the finishing one - could lead to zombie entries if dropped, unless my thinking about things is entirely wrong.

smoogipoo replied:
A start event can't be dropped, as it's `BindUserBeganPlaying(, true)`, right? I'm actually more concerned about event doubling because `BindUserBeganPlaying()` binds first and locks after...
As for the finishing event - I think you're right that there's a possibility it can be dropped - basically a thread race here. Perhaps binding the finishing event first?

Perhaps showcases an advantage with the way multiplayer does things as an alternative - everything's scheduled from there.

Will need to look into this again.

smoogipoo replied:
Should have fixed these two issues - for the first, I'm now locking around the entire block. For the second I added another check in the `userBeganPlaying` callback to make sure the user is actually playing, by retrieving their current state.

bdach reviewed (commented):
everything else looks good to go

peppy reviewed (dismissed):
Looks pretty sane, pending an answer to pending review.

smoogipoo committed:
Merge branch 'master' into fix-initial-spectator-state-callback

smoogipoo committed:
Fix potentially starting play when finished

The UserFinishedPlaying event may trigger before the event is subscribed
to by SpectatorScreen. For such cases, an extra check is done to make
sure the user is _actually_ playing.

smoogipoo committed:
Fix potential doubling of events

peppy committed:
Merge branch 'master' into fix-initial-spectator-state-callback

peppy reviewed (approved):
Not 100% on the safety of this but I can't see any obvious errors.
